### ISSUE TITLE
Feature request: sending midi cc

### ISSUE THREAD

**Issue by upratathlavu** [monocasual/giada#357]
Hi, is there any plan to implement midi cc messages sending? I would like to be able to control my guitar hardware looper (tc electronic ditto x4), but didn't find this functionality. Thank you.

**Comment by monocasual** [comment]
Hi @upratathlavu, thanks for your feedback. So If I understand correctly, you want to send additional messages beyond Note On and Note Off to your looper. I'm not familiar with the X4 Looper, what would you use CC messages for?

**Comment by upratathlavu** [comment]
Hi @monocasual, thanks for reply. Yes, exactly. There are several functions you can operate with midi cc messages on ditto x4, for example starting/stopping loop recording - you can see complete list of them here (there was an official page, but it isn't available at the moment, but the list of messages is basically copy pasted): https://www.elektronauts.com/t/noob-midi-cc-question-regarding-ditto-x4-solved/89599/4.

**Comment by tomek-szczesny** [comment]
Personally I don't see a problem with Giada MIDI channels supporting CC messages along with Note On/ Note Off. I think that the biggest challenge would be to figure out how to represent all that in the editor in a feasible fashion.
This way one could use Giada as a virtual MIDI controller, just like Giada can already be used as a sampler. That could be another cool feature to brag about! ;)

I think that @upratathlavu would also need a MIDI multichannel support to fully control his device, but this is already something under discussion in #252  .

**Comment by monocasual** [comment]
@tomek-szczesny is it weird/unusual to use another widget similar to the Envelope Editor for such things? In other words, a bunch of dots you can drag around, where the X axis is the frame (i.e. position), while the Y axis is the value (from 0 to 127).

**Comment by upratathlavu** [comment]
Thanks :)

**Comment by tomek-szczesny** [comment]
I think that a velocity editor makes a lot of sense for musical notes, but I can hardly imagine similar interface for defining CC messages from 0-127 range, and in virtually random order.
How about a similar editor, but not with a complete 0-127 list, but with a user's selection of CC commands instead? For example, if someone uses only CC 05, 82 and 112, let only these three appear on the list. More might be added later, manually or by recording from a MIDI device.

Also, I feel that CCs should be added to the existing Note On/Off editor, rather than creating a separate one. Why not sending both, after all?

**Comment by upratathlavu** [comment]
I agree with tomek with addition that it would be great to have a way to set cc value by typing not grabbing the fader (as is done in velocity editor). Seq64 has something like this - you can use the same editor for note on/off or cc (or pc), but you cant set cc value by typing it but you at least can se the value (but having possibility to type it would be definitely better). But also, since sometimes you have a control you want to change continually (e. g. volume), not just turn on or off, it would be nice to have some automation graph (as in Ardour).

**Comment by tomek-szczesny** [comment]
Okay, for whatever reason I completely forgot that CC also have nonbinary values :D And I agree this is important, since skilled keyboard players could record notes, volume, pitch bend and whatnot, all at the same time.

It surely would be nice to have it all, but the point is to make it all easily accessible, visible, and not bloated. After all, we all love Giada for its power AND simplicity. I wouldn't resort to tabs or submenus if that's not necessary. Personally I think the key is to smartly hide stuff that is currently not in use and keep the GUI clean, unless this is what some power user actually wants. And do it in a way that any casual user still knows these functionalities are out there to explore.

There is also that old #252, support for MIDI channels (1 - 16), that ALSO must fit into a GUI somehow.

Thanks @upratathlavu for voicing your opinion, perhaps I'll have a look at seq and Ardour and see if they offer any applicable inspiration.

**Comment by tomek-szczesny** [comment]
![obraz](https://user-images.githubusercontent.com/44300715/88480023-affc3c80-cf53-11ea-8840-be778fe6e5cf.png)
Okay, let's think here for a minute.

The way I see it:
1) If loop is shorter than full 32 beats, this ugly grey area is here. If notes were present before shortening loop, they are still there, but not possible to edit (once grabbed with a mouse, they will be moved to active zone).
I think there's no need to restrict users to currently playing loop only. I think a full 32 beats editor should always be present, but with "not playing zone" in a different color.
1b) Action Editor upon opening should be properly sized to whatever is going to appear in it. It may be small enough to hide "Grey Zone" if the currently playing loop is shorter. A horizontal scroll bar is enough.

2) Vertical scroll bar for Note Editor should be on a left hand side. Currently it's far at the end of the whole Grey Zone.
2b) By default this bar should be set somewhere closer to the middle of the range, not the very beginning of the list.

3) Mouse wheel currently zooms time axis. I think that a better use of mouse wheel would be to scroll vertically through MIDI notes or the whole window, depending on current cursor position.

4) There's room for buttons 1-16 above Note Editor - let's use this space for MIDI channel selector. As discussed in #252, a channel shall remember all notes on all channels. One channel may be presented at a time, with notes on other channels in dimmed colors.
4b) Channel 10, [recognized by General MIDI standard as a channel for percussive instruments](https://en.wikipedia.org/wiki/General_MIDI#Percussive), may have more meaningful labels for each sound, like "Kick Drum", "Snare" etc.

5) Below Velocity Editor, a small "+" button could be placed to add CC message timelines. A small menu would pop up, to add either a predefined message timeline, like Balance, Pan, Modulation, Soft Pedal etc, and also "Generic CC", with no value, a binary value, single byte digital, single byte analog or two bytes analog. "Digital" means that user inputs a specific value (a number) for each message, and "Analog" is presented as a graph.
Messages like Pitch Bend / Modulation are well defined within the standard. Modulation is actually a two byte value message with 0x8000 as a default midpoint, and Balance is a single byte parameter. Some others, like Foot Pedal, only accept binary values (MSB from what I understand, so any value from 64 to 127 will do, but it may be a user setting as well). 
All actions recorded from a MIDI controller would of course make the relevant timelines show up.

5b) Instead of a "+" button with a menu, there might be as well a horizontal set of buttons representing each menu entry. Once a timeline is created, its button disappears. It won't take more vertical space than a single button anyway.

6) On "analog" timelines, like Velocity or Panning, dragging each slider is a chore. A "drawing" mouse action would be desired. Still, precise manipulation of each slider should still be a possibility.

7) Under the hood: Most CC messages must not be binded to notes, like Velocity. I think a simple list of time-sorted CC messages in a MIDI channel is a way to go, to be parsed by Action Editor later on.

**Comment by monocasual** [comment]
@tomek-szczesny all your notes regarding the Action Editor improvements definitely make sense, please open a new issue to discuss them. In this one I'll focus on the CC editor implementation.

I pretty much agree with your point 5), a kind of "+" button that creates a new CC message widget. The list of known CC messages can be copy-pasted from the [official list](https://www.midi.org/specifications/item/table-3-control-change-messages-data-bytes-2), plus the generic CC. 

>  Some others, like Foot Pedal, only accept binary values (MSB from what I understand, so any value from 64 to 127 will do, but it may be a user setting as well).

If you really wanna get fancy you can display binary values in the UI. Some DAWs e.g. Ableton Live don't care about it: you always work with an envelope editor that ranges from 0 to 127. From the [Ableton manual page](https://help.ableton.com/hc/en-us/articles/360010389480-Using-MIDI-CC-in-Live#:~:text=MIDI%20CC%20stands%20for%20Control,%2C%20Pitchbend%20and%20Program%20change):

> If a parameter requires an on/off message (eg. CC 64 - Hold/Sustain) then a single breakpoint anywhere between 0-63 is "Off" and 64-127 is "On".

**Comment by tomek-szczesny** [comment]
> If you really wanna get fancy you can display binary values in the UI. Some DAWs e.g. Ableton Live don't care about it: you always work with an envelope editor that ranges from 0 to 127. From the [Ableton manual page](https://help.ableton.com/hc/en-us/articles/360010389480-Using-MIDI-CC-in-Live#:~:text=MIDI%20CC%20stands%20for%20Control,%2C%20Pitchbend%20and%20Program%20change):

I guess you're right. My idea for binary CC was using blocks instead of graph to save some space, but I think this is such a rare use that it's probably not worth the trouble right now.

I was also wrong about my assumptions regarding CC - all of them are single data byte only, and aside from a few predefined by the standard, the rest might as well be a collection of points in 0-127 range. All of them will be handled the same anyway.

**Comment by upratathlavu** [comment]
Just wanted to thank you for adoption of this idea :)
